### PR TITLE
feat(router): get filters for payments

### PR DESCRIPTION
**Pull request by apoorvdixit88**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description
Added API to extract list of filters for payments. These filters can be used to filter payment list on the basis of given constraints.

### Additional Changes

- [x] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?
[
Screenshot 2023-07-04 at 3 29 36 PM
](url)

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/core/payments.rs [hunk 1]
```diff
@@ -1219,6 +1219,39 @@ pub async fn list_payments(
...
+    merchant: domain::MerchantAccount,
+    constraints: api::TimeRange,
+) -> RouterResponse<api::PaymentListFilters> {
+    use crate::types::transformers::ForeignFrom;
+
+    let merchant_id = &merchant.merchant_id;
+    let pi = helpers::filter_by_time_range_constraint(
+        db,
+        &constraints,
+        merchant_id,
```
Comment (lsampras) [inline crates/router/src/core/payments.rs:1234]: `&merchant.merchant_id` ?

File: crates/router/src/core/payments.rs [hunk 2]
```diff
@@ -1219,6 +1219,39 @@ pub async fn list_payments(
...
+        merchant.storage_scheme,
+    )
+    .await
+    .to_not_found_response(errors::ApiErrorResponse::PaymentNotFound)?;
+
+    let filters = db
+        .find_filters_for_payments(
+            &pi,
+            merchant_id,
```
Comment (lsampras) [inline crates/router/src/core/payments.rs:1243]: `&merchant.merchant_id` ?

File: crates/router/src/core/payments.rs [hunk 3]
```diff
@@ -1219,6 +1219,39 @@ pub async fn list_payments(
     ))
 }
 
+#[cfg(feature = "olap")]
+pub async fn filters_for_payments(
+    db: &dyn StorageInterface,
+    merchant: domain::MerchantAccount,
+    constraints: api::TimeRange,
```
Comment (lsampras) [inline crates/router/src/core/payments.rs:1226]:
```suggestion
    time_range: api::TimeRange,
```

File: crates/router/src/core/payments/helpers.rs [hunk 4]
```diff
@@ -1424,6 +1424,19 @@ pub(super) async fn filter_by_constraints(
     Ok(result)
 }
 
+#[cfg(feature = "olap")]
+pub(super) async fn filter_by_time_range_constraint(
+    db: &dyn StorageInterface,
+    constraints: &api::TimeRange,
```
Comment (lsampras) [inline crates/router/src/core/payments/helpers.rs:1430]:
```suggestion
    time_range: &api::TimeRange,
```

File: crates/storage_models/src/query/payment_attempt.rs [hunk 5]
```diff
@@ -174,4 +181,77 @@ impl PaymentAttempt {
...
+    ) -> CustomResult<PaymentListFilters, DatabaseError> {
+        let active_attempts: Vec<String> = pi
+            .iter()
+            .map(|payment_intent| payment_intent.clone().active_attempt_id)
+            .collect();
+
+        let filter = <Self as HasTable>::table()
+            .filter(dsl::merchant_id.eq(merchant_id.to_owned()))
+            .filter(dsl::attempt_id.eq_any(active_attempts))
+            .order(dsl::modified_at.desc());
```
Comment (lsampras) [inline crates/storage_models/src/query/payment_attempt.rs:197]: order not needed probably here

File: crates/router/src/db/payment_intent.rs [hunk 6]
```diff
@@ -304,6 +331,19 @@ mod storage {
                 .map_err(Into::into)
                 .into_report()
         }
+        #[cfg(feature = "olap")]
+        async fn filter_by_time_range_constraints(
+            &self,
+            merchant_id: &str,
+            pc: &api::TimeRange,
```
Comment (lsampras) [inline crates/router/src/db/payment_intent.rs:338]:
```suggestion
            time_range: &api::TimeRange,
```

File: crates/router/src/db/payment_intent.rs [hunk 7]
```diff
@@ -234,6 +242,25 @@ mod storage {
...
                 }
 
+                enums::MerchantStorageScheme::RedisKv => Err(errors::StorageError::KVError.into()),
+            }
+        }
+        #[cfg(feature = "olap")]
+        async fn filter_by_time_range_constraints(
+            &self,
+            merchant_id: &str,
+            pc: &api::TimeRange,
```
Comment (lsampras) [inline crates/router/src/db/payment_intent.rs:252]:
```suggestion
            time_range: &api::TimeRange,
```

File: crates/storage_models/src/query/payment_attempt.rs [hunk 8]
```diff
@@ -174,4 +181,77 @@ impl PaymentAttempt {
         )
         .await
     }
+    pub async fn find_filters_for_payments(
+        conn: &PgPooledConn,
+        pi: &[PaymentIntent],
+        merchant_id: &str,
+    ) -> CustomResult<PaymentListFilters, DatabaseError> {
```
Comment (lsampras) [inline crates/storage_models/src/query/payment_attempt.rs:188]: why not `StorageResult<PaymentListFilters>`

File: crates/storage_models/src/query/payment_attempt.rs [hunk 9]
```diff
@@ -174,4 +181,77 @@ impl PaymentAttempt {
...
+            .distinct()
+            .order_by(dsl::payment_method.asc())
+            .get_results_async::<Option<enums::PaymentMethod>>(conn)
+            .await
+            .into_report()
+            .change_context(DatabaseError::Others)
+            .attach_printable("Error filtering records by payment method")?
+            .into_iter()
+            .flatten()
+            .collect::<Vec<enums::PaymentMethod>>();
```
Comment (lsampras) [inline crates/storage_models/src/query/payment_attempt.rs:246]: use `tokio::join_all` here...

**Review by lsampras — APPROVED** [review]

File: crates/api_models/src/payments.rs [hunk 10]
```diff
@@ -1516,6 +1517,34 @@ pub struct PaymentListResponse {
     pub data: Vec<PaymentsResponse>,
 }
 
+#[derive(Clone, Debug, serde::Serialize, ToSchema)]
+pub struct PaymentListFilters {
```
Comment (Narayanbhat166) [inline crates/api_models/src/payments.rs:1521]: The name seems to indicate that this is a filter, but it seems to be a response based on serde::Serialize derive. Can you change the name accordingly to indicate that it is a response?

File: crates/router/src/core/payments/helpers.rs [hunk 11]
```diff
@@ -1424,6 +1424,19 @@ pub(super) async fn filter_by_constraints(
     Ok(result)
 }
 
+#[cfg(feature = "olap")]
+pub(super) async fn filter_by_time_range_constraint(
```
Comment (Narayanbhat166) [inline crates/router/src/core/payments/helpers.rs:1428]: what is the need of this function? Can't the inner function be called directly?

File: crates/router/src/db/payment_attempt.rs [hunk 12]
```diff
@@ -798,6 +829,27 @@ mod storage {
...
+            storage_scheme: enums::MerchantStorageScheme,
+        ) -> CustomResult<storage_models::payment_attempt::PaymentListFilters, errors::StorageError>
+        {
+            let database_call = || async {
+                let conn = connection::pg_connection_read(self).await?;
+                PaymentAttempt::get_payment_filters(&conn, pi, merchant_id)
+                    .await
+                    .map_err(Into::into)
+                    .into_report()
+            };
```
Comment (Narayanbhat166) [inline crates/router/src/db/payment_attempt.rs:846]: why do we need to do this? wrap the db call in an async block. Will this be used in the `RedisKv` branch?

File: crates/router/src/types/storage/payment_intent.rs [hunk 13]
```diff
@@ -85,4 +91,34 @@ impl PaymentIntentDbExt for PaymentIntent {
...
+
+        filter = filter.filter(dsl::created_at.ge(start_time));
+
+        filter = filter.filter(dsl::created_at.le(end_time));
+
+        filter
+            .get_results_async(conn)
+            .await
+            .into_report()
+            .change_context(errors::DatabaseError::NotFound)
```
Comment (Narayanbhat166) [inline crates/router/src/types/storage/payment_intent.rs:121]: We cannot directly change the context to `NotFound` for all the errors. you can remove this change context

**Review by Narayanbhat166 — CHANGES_REQUESTED** [review]

**Review by Narayanbhat166 — APPROVED** [review]

**Review by lsampras — APPROVED** [review]

File: crates/api_models/src/payments.rs (hunk 10 above)
Comment (Ariessanea1) [reply crates/api_models/src/payments.rs:1602]: Ok what name should it be

**Pull request merged**